modules/events.js

```diff
@@ -32,286 +32,403 @@
+    if (!this._events) {
+        this._events = {};
+    }
+    if (!this._events[type]) {
+        this._events[type] = [];
+    }
+    if (!isArray(this._events[type])) {
+        this._events[type] = [this._events[type]];
+    }
+    return this._events[type];
 };
 
-// Added for CasperJS: filters a value attached to an event
+/**
+ * Trigger filter attached to an event and return value(s).
+ *
+ * @param  String    type      An event name
+ * @param  Array     args...  The rest of arguments passed to fn
+ * @return mixed
+ */
 EventEmitter.prototype.filter = function filter() {
-  var type = arguments[0];
-  if (!this._filters) {
-    this._filters = {};
-    return;
-  }
-
-  var _filter = this._filters[type];
-  if (typeof _filter !== 'function') {
-    return;
-  }
-  return _filter.apply(this, Array.prototype.splice.call(arguments, 1));
+    "use strict";
+    var type = arguments[0], res;
+    if (!this._filters) {
+        this._filters = {};
+    }
+    var _filter = this._filters[type];
+    if (typeof _filter === 'function') {
+        res = _filter.apply(this, Array.prototype.slice.call(arguments, 1));
+    }
+    this.emit(Array.prototype.slice.call(arguments, 0).concat([res]));
+    return res;
 };
 
+/**
+ * Remove all filters attached to events or only filters attached to a specific event.
+ *
+ * @param  String    type      An event name
+ * @return EventEmitter        A reference to the EventEmitter so calls can be chained
+ */
 EventEmitter.prototype.removeAllFilters = function removeAllFilters(type) {
-  if (arguments.length === 0) {
-    this._filters = {};
+    "use strict";
+    if (arguments.length === 0) {
+        this._filters = {};
+        return this;
+    }
+    if (type && this._filters && this._filters[type]) {
+        this._filters[type] = null;
+    }
     return this;
-  }
-  if (type && this._filters && this._filters[type]) {
-    this._filters[type] = null;
-  }
-  return this;
 };
 
+/**
+ * attaches a filter to an event.
+ *
+ * @param  String    type      An event name
+ * @param  function  filterFn  An options callback to apply on event
+ * @return Boolean
+ * @throws {CasperError}       If invokation failed
+ */
 EventEmitter.prototype.setFilter = function setFilter(type, filterFn) {
-  if (!this._filters) {
-    this._filters = {};
-  }
-  if ('function' !== typeof filterFn) {
-    throw new CasperError('setFilter only takes instances of Function');
-  }
-  if (!this._filters[type]) {
-    this._filters[type] = filterFn;
-    return true;
-  }
-  // TODO: process multiple filters? in which order? disallow?
-  return false;
+    "use strict";
+    if (!this._filters) {
+        this._filters = {};
+    }
+    if (typeof filterFn !== 'function') {
+        throw new CasperError('setFilter only takes instances of Function');
+    }
+    if (!this._filters[type]) {
+        this._filters[type] = filterFn;
+        return true;
+    }
+    // TODO: process multiple filters? in which order? disallow?
```

Comment:
how about this todo ?

@casperjs/contributors would you mind to take a look ? :+1: